File: api-factory/af-leader/src/main/java/com/example/afleader/config/CuratorClientConfig.java
package com.example.afleader.config;

import lombok.RequiredArgsConstructor;
import org.apache.curator.RetryPolicy;
import org.apache.curator.framework.CuratorFramework;
import org.apache.curator.framework.CuratorFrameworkFactory;
import org.apache.curator.retry.ExponentialBackoffRetry;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
@RequiredArgsConstructor
public class CuratorClientConfig {

    private final ZookeeperProperties zookeeperProperties;

    @Bean
    public CuratorFramework curatorClient() {
        RetryPolicy retryPolicy = new ExponentialBackoffRetry(zookeeperProperties.getSleepMsBetweenRetry(), zookeeperProperties.getMaxRetries());
        CuratorFramework client = CuratorFrameworkFactory.builder()
                .connectString(zookeeperProperties.getConnectString())
                .connectionTimeoutMs(zookeeperProperties.getConnectionTimeout())
                .sessionTimeoutMs(zookeeperProperties.getSessionTimout())
                .retryPolicy(retryPolicy)
                .namespace(zookeeperProperties.getNamespace())
                .build();
        client.start();
        return client;
    }

}

File: api-factory/af-gateway/src/main/kotlin/com/example/afgateway/config/CuratorClientConfig.kt
package com.example.afgateway.config

import org.apache.curator.framework.CuratorFramework
import org.apache.curator.framework.CuratorFrameworkFactory
import org.apache.curator.retry.ExponentialBackoffRetry
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class CuratorClientConfig(val zookeeperProperties: ZookeeperProperties) {

    @Bean
    fun curatorClient(): CuratorFramework {
        val retryPolicy = ExponentialBackoffRetry(zookeeperProperties.sleepMsBetweenRetry, zookeeperProperties.maxRetries)
        val curatorFramework = CuratorFrameworkFactory.builder()
            .connectString(zookeeperProperties.connectString)
            .connectionTimeoutMs(zookeeperProperties.connectionTimeout)
            .sessionTimeoutMs(zookeeperProperties.sessionTimeout)
            .retryPolicy(retryPolicy)
            .namespace(zookeeperProperties.namespace)
            .build()
        curatorFramework.start()
        return curatorFramework
    }
}